pkg/vmprovider/providers/vsphere/vmprovider_vm_test.go

```diff
@@ -212,6 +214,97 @@ func vmTests() {
 				})
 			})
 
+			Context("ConfigSpec specifies hardware spec", func() {
+				BeforeEach(func() {
+					configSpec = &types.VirtualMachineConfigSpec{
+						Name:     "dummy-VM",
+						NumCPUs:  7,
+						MemoryMB: 5102,
+					}
+				})
+
+				It("CPU and memory from ConfigSpec are ignored", func() {
+					Expect(vm.Status.Phase).To(Equal(vmopv1alpha1.Created))
+
+					vmClass := &vmopv1alpha1.VirtualMachineClass{}
+					Expect(ctx.Client.Get(ctx, client.ObjectKey{Name: vm.Spec.ClassName}, vmClass)).To(Succeed())
+
+					var o mo.VirtualMachine
+					Expect(vcVM.Properties(ctx, vcVM.Reference(), nil, &o)).To(Succeed())
+					Expect(o.Summary.Config.NumCpu).To(BeEquivalentTo(vmClass.Spec.Hardware.Cpus))
+					Expect(o.Summary.Config.NumCpu).To(Not(BeEquivalentTo(configSpec.NumCPUs)))
+					Expect(o.Summary.Config.MemorySizeMB).To(BeEquivalentTo(vmClass.Spec.Hardware.Memory.Value() / 1024 / 1024))
+					Expect(o.Summary.Config.MemorySizeMB).To(Not(BeEquivalentTo(configSpec.MemoryMB)))
+				})
+			})
+
+			Context("VM Class spec CPU reservations are zero and ConfigSpec specifies CPU reservation", func() {
+				BeforeEach(func() {
+					vmClass.Spec.Policies.Resources.Requests.Cpu = resource.MustParse("0")
+					vmClass.Spec.Policies.Resources.Limits.Cpu = resource.MustParse("0")
+
+					// Specify a CPU reservation via ConfigSpec
+					rsv := int64(6)
+					configSpec = &types.VirtualMachineConfigSpec{
+						Name: "dummy-VM",
+						CpuAllocation: &types.ResourceAllocationInfo{
+							Reservation: &rsv,
+						},
+					}
+				})
+
+				It("VM gets CPU reservation from ConfigSpec", func() {
+					Expect(vm.Status.Phase).To(Equal(vmopv1alpha1.Created))
+
+					vmClass := &vmopv1alpha1.VirtualMachineClass{}
+					Expect(ctx.Client.Get(ctx, client.ObjectKey{Name: vm.Spec.ClassName}, vmClass)).To(Succeed())
+
+					var o mo.VirtualMachine
+					Expect(vcVM.Properties(ctx, vcVM.Reference(), nil, &o)).To(Succeed())
+
+					// Freq used by vcsim
+					const freq = 2294
+
+					reservation := o.Config.CpuAllocation.Reservation
+					vmClassCPURes := virtualmachine.CPUQuantityToMhz(vmClass.Spec.Policies.Resources.Requests.Memory, freq)
+					Expect(reservation).ToNot(BeNil())
+					Expect(reservation).To(Not(BeEquivalentTo(vmClassCPURes)))
```

Comment:
Nit here & a few other places: our usual pattern is `ToNot()` instead

Reply:
Missed it.  Will take care of it in a folllowup.